File: spark-study-java/src/main/java/com/zhouq/spark/sql/JSONDataSource.java
package com.zhouq.spark.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;

/**
 * JsonData 数据源综合示例
 * <p>
 * Create by zhouq on 2019/9/9
 */
public class JSONDataSource {
    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("JSONDataSource").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        SQLContext sqlContext = new SQLContext(sc);


        DataFrame students_scoreDF = sqlContext.read().json("H:\\bigdatatest\\spark\\sql\\students_score.json");

        students_scoreDF.registerTempTable("students_score");

        // 针对学生基本信息DataFrame，注册临时表，然后查询分数大于80分的学生的基本信息
        DataFrame goodStudentScoresDF = sqlContext.sql("select name,score from students_score where score > 80");

        List<String> goodStudentNames = goodStudentScoresDF.javaRDD().map(new Function<Row, String>() {
            @Override
            public String call(Row row) throws Exception {
                return row.getString(0);
            }
        }).collect();

        //
        List<String> studentInfoJSONs = new ArrayList<String>();

        studentInfoJSONs.add("{\"name\":\"Leo\", \"age\":18}");
        studentInfoJSONs.add("{\"name\":\"Marry\", \"age\":17}");
        studentInfoJSONs.add("{\"name\":\"Jack\", \"age\":19}");

        JavaRDD<String> studentInfoJSONsRDD = sc.parallelize(studentInfoJSONs);

        DataFrame studentsInfoDF = sqlContext.read().json(studentInfoJSONsRDD);

        studentsInfoDF.registerTempTable("students_Info");

        String sql = "select name,age from students_Info where name in (";

        for (int i = 0; i < goodStudentNames.size(); i++) {
            sql += "'" + goodStudentNames.get(i) + "'";
            if (i < goodStudentNames.size() - 1) {
                sql += ",";
            }
        }
        sql += ")";

        DataFrame goodStudentsInfoDF = sqlContext.sql(sql);

        // 然后将两份数据的DataFrame，转换为JavaPairRDD，执行join transformation

        JavaPairRDD<String, Tuple2<Integer, Integer>> goodStudentsRDD =
                // name,age
                goodStudentsInfoDF.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
                    @Override
                    public Tuple2<String, Integer> call(Row row) throws Exception {
                        return new Tuple2<String, Integer>(row.getString(0), Integer.valueOf(String.valueOf(row.getLong(1))));
                    }
                }).join(
                        // name,score
                        goodStudentScoresDF.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
                            @Override
                            public Tuple2<String, Integer> call(Row row) throws Exception {
                                return new Tuple2<>(row.getString(0), Integer.valueOf(String.valueOf(row.getLong(1))));
                            }
                        }));


        // 然后将封装在RDD中的好学生的全部信息，转换为一个JavaRDD<Row>的格式
        JavaRDD<Row> goodStudentRowsRDD = goodStudentsRDD.map(new Function<Tuple2<String, Tuple2<Integer, Integer>>, Row>() {
            @Override
            public Row call(Tuple2<String, Tuple2<Integer, Integer>> tuple) throws Exception {

                return RowFactory.create(tuple._1, tuple._2._1, tuple._2._2);
            }
        });

        //创建一份元数据信息

        List<StructField> structFields = new ArrayList<StructField>();
        structFields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
        structFields.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));
        structFields.add(DataTypes.createStructField("score", DataTypes.IntegerType, true));

        StructType structType = DataTypes.createStructType(structFields);

        DataFrame goodStudentsDF = sqlContext.createDataFrame(goodStudentRowsRDD, structType);
        // 将好学生的全部信息保存到一个json文件中去
        goodStudentsDF.write().format("json").save("H:\\bigdatatest\\spark\\sql\\out\\good-students");

        sc.stop();
    }
}

File: spark-study-scala/src/main/java/com/zhouq/spark/sql/JSONDataSource.scala
package com.zhouq.spark.sql

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.types.{IntegerType, LongType, StringType, StructField, StructType}
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.{DataFrame, Row, SQLContext}

/**
  * Create by zhouq on 2019/9/10
  *
  */
object JSONDataSource {
  def main(args: Array[String]): Unit = {
    val conf: SparkConf = new SparkConf().setMaster("local").setAppName("JSONDataSource")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)

    val studentSourceDF: DataFrame = sqlContext.read.json("H:\\bigdatatest\\spark\\sql\\students_score.json")

    // 注册
    studentSourceDF.registerTempTable("students_score")

    // 查询大于80的 学生
    val goodStudentScoreDF: DataFrame = sqlContext.sql("select name,score from students_score where score > 80")
    // 提取学生名称列表
    val goodStudentNames: Array[Any] = goodStudentScoreDF.rdd.map(row => row(0)).collect()

    // 处理学生信息数据
    val studentInfoJsons = Array("{\"name\":\"Leo\", \"age\":18}", "{\"name\":\"Marry\", \"age\":17}", "{\"name\":\"Jack\", \"age\":19}")
    // 转换为 RDD
    val studentInfoJSONSRDD: RDD[String] = sc.parallelize(studentInfoJsons, 1)

    // 转换为 DF
    val studentInfoJsonDF: DataFrame = sqlContext.read.json(studentInfoJSONSRDD)
    // 注册临时表
    studentInfoJsonDF.registerTempTable("student_info")

    // 查询分数大于80 分的学生的基本信息
    var sql = "select name,age from  student_info where name in ("

    for (i <- 0 until goodStudentNames.length) {
      sql += "'" + goodStudentNames(i) + "'"
      if (i < goodStudentNames.length - 1) {
        sql += ","
      }
    }
    sql += ")"

    val goodStudentInfoDF: DataFrame = sqlContext.sql(sql)

    // 将分数大于80分的学生信息跟成绩信息进行 join

    val goodStudentsRDD: RDD[(String, (Long, Long))] = goodStudentScoreDF.rdd
      .map(
        row => (row.getAs[String]("name"), row.getAs[Long]("score"))
      )
      .join(
        goodStudentInfoDF.rdd
          .map(
            row => (row.getAs[String]("name"), row.getAs[Long]("age"))
          )
      )

    // 将RDD 转换为 DataFrame
    val goodStudentRowsRDD: RDD[Row] = goodStudentsRDD.map(info => Row(info._1, info._2._1.toInt, info._2._2.toInt))

    val structType = StructType(Array(
      StructField("name", StringType, true),
      StructField("score", IntegerType, true),
      StructField("age", IntegerType, true)
    ))

    val goodStudentsDF: DataFrame = sqlContext.createDataFrame(goodStudentRowsRDD, structType)

    // 将DF 中的数据保存为 json 文件
    goodStudentsDF.write.format("json").save("H:\\bigdatatest\\spark\\sql\\out\\good-students-scala")

    sc.stop()

  }

}
